controllers/configurationpolicy_controller.go

```diff
@@ -364,6 +361,103 @@ func (r *ConfigurationPolicyReconciler) handleObjectTemplates(plc policyv1.Confi
+	// - Object namespace, if specified
+	// - Whether object is namespaced
+	//
+	// Also determine whether a namespace query is needed for this policy:
+	//   If a namespaceSelector is present AND objects are namespaced without a namespace specified
+	templateObjs := make([]struct {
+		name         string
+		namespace    string
+		isNamespaced bool
+	}, len(plc.Spec.ObjectTemplates))
+	selectedNamespaces := []string{}
+	queryNamespaces := false
+
+	for idx, objectT := range plc.Spec.ObjectTemplates {
+		var blob interface{}
+		if jsonErr := json.Unmarshal(objectT.ObjectDefinition.Raw, &blob); jsonErr != nil {
+			log.Error(jsonErr, "Could not unmarshal data from JSON")
+
+			return
+		}
+
+		var unstruct unstructured.Unstructured
+		unstruct.Object = make(map[string]interface{})
+		unstruct.Object = blob.(map[string]interface{})
+
+		templateObjs[idx].isNamespaced = r.isObjectNamespaced(&unstruct, true)
+
+		if md, ok := unstruct.Object["metadata"]; ok {
+			metadata := md.(map[string]interface{})
+
+			if objNs, ok := metadata["namespace"]; ok {
+				templateObjs[idx].namespace = objNs.(string)
+			} else if templateObjs[idx].isNamespaced {
+				log.V(2).Info(
+					"metadata.namespace not found or could not be retrieved from the object template", "index", idx,
+				)
+			}
+
+			if objName, ok := metadata["name"]; ok {
+				templateObjs[idx].name = objName.(string)
+			} else {
+				log.V(2).Info(
+					"metadata.name not found or could not be retrieved from the object template", "index", idx,
+				)
+			}
+		}
+
+		if templateObjs[idx].isNamespaced && templateObjs[idx].namespace == "" {
+			queryNamespaces = true
+		}
+	}
+
+	// If required, query for namespaces specified in NamespaceSelector for objects to use
+	if queryNamespaces {
+		// Retrieve the namespaces based on filters in NamespaceSelector
+		selector := plc.Spec.NamespaceSelector
+		// If MatchLabels/MatchExpressions/Include were not provided, return no namespaces
+		if selector.MatchLabels == nil && selector.MatchExpressions == nil && len(selector.Include) == 0 {
+			log.Info("NamespaceSelector is empty. Skipping namespace retrieval.")
+		} else {
+			// If an error occurred in the NamespaceSelector, update the policy status and abort
+			var err error
+			selectedNamespaces, err = common.GetSelectedNamespaces(selector)
+			if err != nil {
+				log.Error(
+					err, "Error filtering namespaces with provided NamespaceSelector",
+					"namespaceSelector", fmt.Sprintf("%+v", selector))
+
+				reason := "NamespaceSelector error"
+				msg := fmt.Sprintf(
+					"Error filtering namespaces with provided NamespaceSelector: %s", err.Error())
+				statusChanged := addConditionToStatus(&plc, 0, false, reason, msg)
+				if statusChanged {
+					parentStatusUpdateNeeded = true
+
+					r.Recorder.Event(
+						&plc,
+						eventWarning,
+						fmt.Sprintf(plcFmtStr, plc.GetName()),
+						convertPolicyStatusToString(&plc),
+					)
+				}
+
+				r.addForUpdate(&plc, statusChanged)
+
+				return
```

Comment:
You'll need to also call `r.checkRelatedAndUpdate` before returning for the status update to take effect.

Reply:
It's probably okay actually to just call `addForUpdate`.

Reply:
I think you're right, though--it's probably clearer to wipe out the related resources if the policy/namespaceSelector is invalidated. I've moved the logic up so that it applies to `remediationAction` also.